pkg/cloud/aws/actuators/cluster/actuator.go

```diff
@@ -64,14 +64,48 @@ func (a *Actuator) Reconcile(cluster *clusterv1.Cluster) error {
 	elbsvc := elb.NewService(scope)
 
 	// Store some config parameters in the status.
-	if len(scope.ClusterConfig.CACertificate) == 0 {
-		caCert, caKey, err := certificates.NewCertificateAuthority()
+	if !scope.ClusterConfig.CAKeyPair.HasCertAndKey() {
+		caCert, caKey, err := actuators.GetOrGenerateKeyPair(&scope.ClusterConfig.CAKeyPair, "cluster-ca")
 		if err != nil {
 			return errors.Wrap(err, "Failed to generate a CA for the control plane")
 		}
+		scope.ClusterConfig.CAKeyPair = v1alpha1.KeyPair{
+			Cert: caCert,
+			Key:  caKey,
+		}
+	}
+
+	if !scope.ClusterConfig.EtcdCAKeyPair.HasCertAndKey() {
+		etcdCACert, etcdCAKey, err := actuators.GetOrGenerateKeyPair(&scope.ClusterConfig.EtcdCAKeyPair, "etcd-ca")
+		if err != nil {
+			return errors.Wrapf(err, "failed to get or generate etcd CA cert")
+		}
+		scope.ClusterConfig.EtcdCAKeyPair = v1alpha1.KeyPair{
+			Cert: etcdCACert,
+			Key:  etcdCAKey,
+		}
+	}
 
-		scope.ClusterConfig.CACertificate = certificates.EncodeCertPEM(caCert)
-		scope.ClusterConfig.CAPrivateKey = certificates.EncodePrivateKeyPEM(caKey)
+	if !scope.ClusterConfig.FrontProxyCAKeyPair.HasCertAndKey() {
+		fpCACert, fpCAKey, err := actuators.GetOrGenerateKeyPair(&scope.ClusterConfig.FrontProxyCAKeyPair, "front-proxy-ca")
+		if err != nil {
+			return errors.Wrapf(err, "failed to get or generate front-proxy CA cert")
+		}
+		scope.ClusterConfig.FrontProxyCAKeyPair = v1alpha1.KeyPair{
+			Cert: fpCACert,
+			Key:  fpCAKey,
+		}
+	}
+
+	if !scope.ClusterConfig.SAKeyPair.HasCertAndKey() {
+		saPub, saKey, err := actuators.GetOrGenerateKeyPair(&scope.ClusterConfig.SAKeyPair, "service-account")
+		if err != nil {
+			return errors.Wrapf(err, "failed to get or generate service-account certificates")
+		}
+		scope.ClusterConfig.SAKeyPair = v1alpha1.KeyPair{
+			Cert: saPub,
+			Key:  saKey,
+		}
```

Comment:
There is a lot of repetition unfortunately, is there a way to use a generic for loop and wrap the above into its own defaulting function?

Reply:
This is wrapped up into this ticket https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/537

I'd rather move this forward than make it perfect